Azure/sonic-buildimage
# Way to skip over TACACS when there is system level error.

**- Why I did it**
Fact: With TACACS configured, only remote users could login. 
When there is a serious system level error, for example Hard-disk becoming read-only, no new users could login, as that requires passwd and other files update along with creation of home dir. The "admin" user would have all entries pre-created, but TACACS would block local user login.

**- How I did it**
Added a script to run at the top of PAM stack, when login=tacacs+. This script would check the system and if there is any serious error, it would return system_err and the corresponding action is to skip the TACACS servers to land in pam_unix.so which would authenticate local users

**- How to verify it**
As it is tough to simulate read-only Hard disk, I hacked the check_system script to return "system_err" if a file /var/log/read_only exists. 
0) Update with following, if you need to test w/o upgrading image
    sudo cp src/sonic-host-services-data/templates/common-auth-sonic.j2 /usr/share/sonic/templates/common-auth-sonic.j2; cp 
    sudo cp src/sonic-host-services/scripts/check_system /usr/local/bin
    sudo chmod +x src/sonic-host-services/scripts/check_system
1) Configure TACACS with fail through not enabled
2) Try login as admin
    Expect: Fail
    Note: If you are running ssh, do Ctrl-C and get out
3) touch /var/log/read_only
4) Try login as admin
     Expect: Success

**- Which release branch to backport (provide reason below if selected)**



- [x ] 201911
- [x ] 202006
- [x ] 202012

**- Description for the changelog**



**- A picture of a cute animal (not mandatory but encouraged)**
